File: libsrc/pylith/fekernels/PointForce.cc
/* -*- C++ -*-
 *
 * ----------------------------------------------------------------------
 *
 * Brad T. Aagaard, U.S. Geological Survey
 * Charles A. Williams, GNS Science
 * Matthew G. Knepley, University of Chicago
 *
 * This code was developed as part of the Computational Infrastructure
 * for Geodynamics (http:*geodynamics.org).
 *
 * Copyright (c) 2010-2015 University of California, Davis
 *
 * See COPYING for license information.
 *
 * ----------------------------------------------------------------------
 */

#include <portinfo>

#include "pylith/fekernels/PointForce.hh"

#include <cassert> // USES assert()
#include <iostream> // debugging.

// ----------------------------------------------------------------------
// f0 entry function for displacement equation: f0u = \dot{u}.
void
pylith::fekernels::PointForce::f0u(const PylithInt dim,
                                   const PylithInt numS,
                                   const PylithInt numA,
                                   const PylithInt sOff[],
                                   const PylithInt sOff_x[],
                                   const PylithScalar s[],
                                   const PylithScalar s_t[],
                                   const PylithScalar s_x[],
                                   const PylithInt aOff[],
                                   const PylithInt aOff_x[],
                                   const PylithScalar a[],
                                   const PylithScalar a_t[],
                                   const PylithScalar a_x[],
                                   const PylithReal t,
                                   const PylithScalar x[],
                                   const PylithInt numConstants,
                                   const PylithScalar constants[],
                                   PylithScalar f0[]) {
    assert(sOff);
    assert(s);
    assert(s_t);
    assert(f0);

    // Incoming re-packed solution field.
    const PylithInt i_displacement = 0;

    // Incoming re-packed auxiliary field.
    const PylithInt i_pointForce = 0;

    const PylithScalar *pointForce = &a[aOff[i_pointForce]];

    for (PylithInt i = 0; i < dim; ++i) {
        f0[i] += pointForce[i];
    } // for
}


// End of file

File: pylith/sources/PointForce.py
# ----------------------------------------------------------------------
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University of Chicago
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2016 University of California, Davis
#
# See COPYING for license information.
#
# ----------------------------------------------------------------------
#
# @file pylith/sources/PointForce.py
#
# @brief Python object for solving the pointforce equation.
#
# Factory: source

from pylith.sources.RickerFunction import RickerFunction
from .Source import Source
from .sources import PointForce as ModulePointForce


class PointForce(Source, ModulePointForce):
    """Python source property manager.

    FACTORY: source
    """

    import pythia.pyre.inventory

    source_time_function = pythia.pyre.inventory.facility("source_time_function", family="pointforce_sourcetimefunction", factory=RickerFunction)
    source_time_function.meta['tip'] = "Source time function for point force."

    # PUBLIC METHODS /////////////////////////////////////////////////////

    def __init__(self, name="pointforce"):
        """Constructor.
        """
        Source.__init__(self, name)
        return

    def _defaults(self):
        from .AuxSubfieldsPointForce import AuxSubfieldsPointForce
        self.auxiliarySubfields = AuxSubfieldsPointForce("auxiliary_subfields")


    def preinitialize(self, problem):
        """Setup source.
        """
        Source.preinitialize(self, problem)


        return

    def _createModuleObj(self):
        """Create handle to C++ PointForce.
        """
        ModulePointForce.__init__(self)
        return


# Factories

def source():
    """Factory associated with PointForce.
    """
    return PointForce()


# End of file
